# Calculating Option Greeks

This notebook shows how to calculate Option Greeks using the pathwise estimation method as well as the likelihood ratio method.

While understanding how to price options is important, it is equally if not more important to understand how the price of an option changes with respect to its underlying inputs.

Recall that the price of an option, under a Black Scholes framework, is dependent on five parameters:

1. The stock price
2. The strike price
3. The time until maturity
4. The volatility of the stock
5. The interest rate

Under a Black Scholes framework, we can compute the partial derivatives with respect to four of these inputs (since strike price is a constant, it does not make sense to compute a derivative with respect to it.) These are commonly referred to as the Greeks, as the names of (some of) these partial derivatives are based on Greek letters. The ones that are not are named to "sound like" Greek letters, but no actual greek letter exists.

### Black Scholes

The Greeks can be computed analytically as shown above. We start with the analytical form of the BS equation:

$ C(S_t, t) $ is the call price and $P(S_t, t) $ is the put price. I will also denote $V$ as a generic option value function, for either a Call or a Put.

$ \begin{aligned}C(S_{t},t)&=N(d_{1})S_{t}-N(d_{2})PV(K)\\d_{1}&={\frac {1}{\sigma {\sqrt {T-t}}}}\left[\ln \left({\frac {S_{t}}{K}}\right)+\left(r+{\frac {\sigma ^{2}}{2}}\right)(T-t)\right]\\d_{2}&=d_{1}-\sigma {\sqrt {T-t}}\\PV(K)&=Ke^{-r(T-t)}\end{aligned}  $

 The price of a corresponding put option based on put–call parity is:
 
$ \begin{aligned}P(S_{t},t)&=Ke^{-r(T-t)}-S_{t}+C(S_{t},t)\\&=N(-d_{2})Ke^{-r(T-t)}-N(-d_{1})S_{t}\end{aligned} $


### The Greeks (Definitions)

Analytical computations/proofs may be added later, but can be found from many reference books.

#### Delta

Delta is defined as the partial derivative of the option value with respect to the underlying.

$Delta = \Delta = \frac{\partial V}{\partial S} $

#### Gamma

Gamma is defined as the partial derivative of an option's delta with respect to the underlying. Thus it is the second derivative of the option value with respect to the underlying.

$Gamma = \Gamma = \frac{\partial \Delta}{\partial S}  = \frac{\partial ^2 V}{\partial S^2}$

#### Vega

Vega is defined as the partial derivative of the option price with respect to the volatility.

$Vega = \mathcal{V} = \frac{\partial V}{\partial \sigma} $

#### Theta

Theta is defined as the partial derivative of the option price with respect to time.

$Theta = \Theta = \frac{\partial V}{\partial t} $

Note that sometimes it is defined as $-\frac{\partial V}{\partial \tau} $. Since $\tau = T - t$, these derivatives are consistent. (Recall that $T$ is the maturity time, $t$ is the variable of time, and $\tau$ measures the time remaining until maturity).


Less important Greeks:

#### Rho

Rho is defined as the partial derivative of the option value with respect to the (risk free) interest rate.
$Rho = \rho = \frac{ \partial V}{\partial r} $


#### Vomma

Vomma is the partial derivative of an option's Vega with respect to volatility. Thus it is the second order partial derivative of the option price function with respect to volatility.

$Vanna = \frac{ \partial \mathcal{V}}{\partial \sigma} = \frac{\partial ^2 V}{\partial \sigma^2}$

#### Vanna

Vanna is the partial derviative of the Vega with respect to the spot price, or the partial derivative of the Delta with respect to volatility. It is thus the 2nd order cross partial of the option value with respect to the spot price and the volatility.

$Vanna = \frac{ \partial \Delta}{\partial \sigma} = \frac{\partial ^2 V}{\partial S \partial \sigma}$

#### Veta
Veta is the partial derviative of the Vega with respect to time, or the partial derivative of the Theta with respect to volatility. It is thus the 2nd order cross partial of the option value with respect to the time and the volatility. It measures how the Vega is changing with the passage of time
$Vanna = \frac{ \partial \mathcal{V}}{\partial \tau} = \frac{\partial ^2 V}{\partial \tau \partial \sigma}$


#### Charm

$Vanna = \frac{ \partial \mathcal{V}}{\partial \sigma} = \frac{\partial ^2 V}{\partial \sigma^2}$


Reference: https://en.wikipedia.org/wiki/Greeks_(finance)

### Finite Differencing

Finite differencing is the simplest way to compute derivatives. It is a numerical estimation method for computing the derivative of a function. It can be applied to many "well behaving" functions. We will focus on two, the forward finite difference, and the central finite difference.

Let use define the following variables:
 $n$ is the number of replications/simulations,
 $h$ is the increment size, or the small perturbation,
$Y_i(\theta)$ is a random variable that is dependent on the parameter $\theta$. Note that $Y_i's$ are i.i.d.

$\alpha (\theta) = \mathbb{E}[Y(\theta)]$

The quantity we are interested in estimating is $\alpha ^\prime (\theta)$, i.e. the derivative of $\alpha$.

In the context of options, $Y(\theta)$ will usually be the discounted payoff of an option, $\alpha$ the price of the option, and $\theta$ the individual variable we are singling out to perform sensitivity analysis on. If we take $\theta$ to be $S$, we will be estimating the delta. If we take $\theta$ to be $\sigma$, then we will be estimating the Vega. Note that the $\theta$ here is simply to define the equations. This is **not** the options Theta.

The forward finite difference is defined as:

$\Delta_F(n,h) = \frac{\bar{Y}_n(\theta+h) - \bar{Y}_n(\theta)}{h}$

Now let's take the expectation of this. On the right hand side, we have, by the linearity property of expectations:

$ \frac{\mathbb{E}[\bar{Y}_n(\theta+h)] - \mathbb{E}[\bar{Y}_n(\theta)]}{h}$

But based on our definition above, we can simply rewrite this as:

$\frac{\alpha(\theta+h) - \alpha(\theta)}{h}$


We know from the Taylor series expansion, that if $\alpha$ is a twice continuously differentiable function, (which it is in our case), then

$\alpha(\theta + h) = \alpha(\theta) + \alpha ^\prime(\theta)h + \frac{1}{2} \alpha ^{\prime\prime}(\xi)h^2$

where $\xi$ is a small value that lies in $(\theta, \theta + h)$

Therefore, if we plug this result in to our equation for the expectation, we have:

$\mathbb{E}[\Delta_F(n,h)] = \alpha ^\prime(\theta) + \frac{1}{2} \alpha ^{\prime\prime}(\xi)h$

But recall our goal was to compute $\alpha ^\prime(\theta)$. Since this is not our expected value, our estimator has a bias of $\frac{1}{2} \alpha ^{\prime\prime}(\xi)h = \frac{1}{2} \alpha ^{\prime\prime}(\theta)h + o(h)$ where $o(h)$ is notation for a deterministic function $\psi$ with the property that $\lim_{h \to 0} \frac{\psi}{h} \to 0$

The important thing to know is that the forward finite difference is biased and the bias is proprotional to the value of the second derivative. For example, if we compute a forward finite difference for an Option's Delta, it will be biased in proportion to its Gamma.

Alternatively we can use the central finite difference, defined as:

$\Delta_C(n,h) = \frac{\bar{Y}_n(\theta+h) - \bar{Y}_n(\theta - h)}{2h}$

Using results from above, we can easily replace the "+" with a "-" to compute the 2nd term in this expression, and then easily show that:

$\mathbb{E}[\Delta_C(n,h)] = \frac{\alpha(\theta + h) + \alpha(\theta - h)}{2h} = (2\alpha^\prime(\theta)h)/(2h) = \alpha ^\prime(\theta) + o(h)$.

Just as above, the $o(h)$ term can be ignored. For all intents and purposes, the central finite difference provides us with an unbiased estimator.

Note the tradeoff: We always simulate at the value $\theta$. With forward finite differences, we perform simply one additional simulation, at $\theta + h$. With central finite differences, we perform a total of two additional simulations. 

Usually, the additional simulation is worth it as we value the reduction in bias.

#### Variance of Finite Differences

I have ommitted the relevant proof here as this notebook is already getting rather long, but we do not want to simply take h as small as possible. We want to minimize the mean square error, which is the sum of the variance and the square of the bias. If we use central diferences, the bias is zero, so we only need to minimize the variance. It can be shown that 
$h$ should be of the order $n^{1/6}$ to minimize the MSE around the order of $n^{-2/3}$.

### Simple Analysis

Let us first begin with some simple analysis. We will construct our standard Monte Carlo set up  to price options. We will choose the following parameters:

$S_0 = 1200, r = 0.03, q = 0.015, T = 1, \sigma = 0.3,$ and $K = 1150$

We will simulate at two points in time for simplicity. Once at $t_1 = 0.5$ and another at $T=1$

Let us estimate for a European put the Delta, Gamma, and Vega. Since this is a European option, we can compute the analytical values from our Black Scholes functions as well to compare. We will see later however that our estimation methods can work for exotic options, while these analytical computations will fail.

We will assume 1 million sample paths. We will also introduce two parameters, delS and delSig, which represent the small increments we use to perform finite differencing to estimate the Greeks.

In [7]:
import numpy as np
from scipy.stats import norm

## Define Parameters
s0 = 1200
K = 1150

sig = 0.3
T = 1
#t1 = 0.5
r = 0.03
q = 0.015
#lmbda = 0.9

# optimal h: 1e6**(-1/6)
# This equals 0.1
# We will use this for delta.
# For comparison, we will use a smaller delta for the vega
# These are the "h" values

delS = 0.1
delSig = 0.001

#1 million paths
nN = int(1e6)


## Analytical BS values for reference

d1 = (np.log(s0/K) + (r - q + sig**2/2)*T)/(sig*np.sqrt(T))
d2 = d1 - sig*np.sqrt(T)

delC =  np.exp(-q*T) * norm.cdf(d1)
delP = -np.exp(-q*T) * norm.cdf(-d1)

gammaC = np.exp(-q*T)*norm.pdf(d1)/(s0*sig*np.sqrt(T))
# gammaP = gammaC

vegaC = s0*np.sqrt(T)*np.exp(-q*T)*norm.pdf(d1)
# vegaP = vegaC
print("The analytically computed values from the closed form solution of the Black Scholes equation are: ")
print(f"""delta Call: {delC}
delta Put: {delP}
gamma : {gammaC}
vega : {vegaC}
""")

The analytically computed values from the closed form solution of the Black Scholes equation are: 
delta Call: 0.6243382342789829
delta Put: -0.3607737053240797
gamma : 0.001029709521832556
vega : 444.8345134316643



### Common Random Numbers

Before we continue with the code, it is important to explain what common random numbers are and why they are vital here.

Common random numbers is (an implicit) variance reduction technique. That is to say, it results in lowered variance by the very nature of keeping the random numbers *common* between the samples generated. However, unlike other variance reduction techniques, it does not introduce any additional statistics or variables. It can be combined with other *explicit* variance reduction techniques as well, if appropriate.

Common random numbers is unrelated to the random number sequence generation. Built in generators in standard packages are psuedo-random, but are sufficiently random for our purposes. They have a less uniform dispersion compared to nonrandom sequences used for simulations, such as the Sobol sequence. You can read more about that here: https://en.wikipedia.org/wiki/Sobol_sequence

The code below computes the Option's delta. There is a flag present that affects a simple change in an if/else block regarding what random numbers to use. If CRN is true, we use the same random numbers for all 3 paths for central finite difference approximations.

In [44]:

## 1. Simple  Simulation and CRN

def compute( CRN ):

      ## Simulate three times, one for the stock price process, one minus delS, one plus delS
      np.random.seed(12345)
      z1 = np.random.randn(nN)
      z2 = np.random.randn(nN)
      z3 = np.random.randn(nN)

      if(CRN):
            STU  = (s0 + delS)*np.exp((r-q-sig**2/2)*T + sig*np.sqrt(T)*z1)
            ST   = s0         *np.exp((r-q-sig**2/2)*T + sig*np.sqrt(T)*z1)
            STD  = (s0 - delS)*np.exp((r-q-sig**2/2)*T + sig*np.sqrt(T)*z1)
      else:
            STU  = (s0 + delS)*np.exp((r-q-sig**2/2)*T + sig*np.sqrt(T)*z1)
            ST   = s0         *np.exp((r-q-sig**2/2)*T + sig*np.sqrt(T)*z2)
            STD  = (s0 - delS)*np.exp((r-q-sig**2/2)*T + sig*np.sqrt(T)*z3)

      #compute payoffs using the central point
      tmpC = np.exp(-r*T)*np.maximum(ST-K, 0) 
      tmpP = np.exp(-r*T)*np.maximum(K-ST, 0)

      #compute payoffs using the forward point
      tmpUC = np.exp(-r*T)*np.maximum(STU - K,0) #CALL
      tmpUP = np.exp(-r*T)*np.maximum(K - STU,0) #PUT

      #compute payoffs using the backward point
      tmpDC = np.exp(-r*T)*np.maximum(STD - K,0) #CALL
      tmpDP = np.exp(-r*T)*np.maximum(K - STD,0) #PUT

      #Deltas (Greeks), forward and central estimators

      delC_Fwd = np.mean((tmpUC - tmpC))/delS
      delC_Central = np.mean((tmpUC - tmpDC))/(2*delS)

      delP_Cwd = np.mean((tmpUP - tmpP))/delS
      delP_Central = np.mean((tmpUP - tmpDP))/(2*delS)
      
      gam_Central  = np.mean((tmpUC - 2*tmpC + tmpDC))/(delS**2)

      print(f"""
      The forward finite difference estimator for the Delta is:
            Call: {delC_Fwd:.4f}
            Put: {delP_Cwd:.4f}
      The central finite difference estimator for the Delta is:
            Call: {delC_Central:.4f}
            Put: {delP_Central:.4f}
      The central finite difference estimator for the Gamma is:
            {gam_Central:.4f} (same for call and put)
      """)

print("When not using Common Random numbers:")
compute(CRN = False)

When not using Common Random numbers:

      The forward finite difference estimator for the Delta is:
            Call: 2.3159
            Put: -1.6057
      The central finite difference estimator for the Delta is:
            Call: 2.2810
            Put: -1.7691
      The central finite difference estimator for the Gamma is:
            0.6972 (same for call and put)
      


In [45]:
print("When using CRN:")
compute(CRN = True)

When using CRN:

      The forward finite difference estimator for the Delta is:
            Call: 0.6251
            Put: -0.3604
      The central finite difference estimator for the Delta is:
            Call: 0.6251
            Put: -0.3605
      The central finite difference estimator for the Gamma is:
            0.0011 (same for call and put)
      


Recall from above that the BSM equation yielded an analytical call delta of 0.624 and a Put Delta of -0.361. As we can see, using common random numbers we get results that make sense. Not using them lead to garbage results

The moral of the story here is that when doing sensitivity analysis, it is mandatory to use CRN because we need to perform a sensitivity analysis on each of the unique paths that generated the original sample (S not shifted up or down).

### Pathwise Estimation

Pathwise estimation, which is also sometimes referred to as infinitesimal perturbation analysis (IPA), is a smarter method of estimation. Recall our definition at the start of the our finite differences discussion:

$\alpha (\theta) = \mathbb{E}[Y(\theta)]$

Pathwise estimation relies on interchanging the order of differentiation and integration to obtain

$\alpha ^\prime (\theta) = \frac{d}{d\theta}\mathbb{E}[Y(\theta)] = \mathbb{E} \left[ \frac{dY(\theta)}{d\theta} \right]$

Research has been done regarding when the interchange is valid. In summary, if, with probability 1, $Y$ is a continuous function of $\theta$, then the interchange is valid.

This means, that if we take $Y(\theta) = e^{-rT}[S(T) - K]^+$, i.e. the payoff of a European call, $Y$ is continuous and so the interchange applies. Remember, the payoff function has a kink at $S_T = K$, but it is continuous (recall the definitions of continuity and limits from calculus if this is confusing). The derivative of this payoff function does not exist at $S_T = K$, thus, this payoff function is continuous but not differentiable.

It is important to note, that due to this condition, the pathwise estimation method does not always work. Consider a Digital option whose payoff is $Y = e^{-rT} \mathbb{I}_{S_T > K}$. This payoff function is *discontinuous* with a discontinuity at $S_T = K$ and so pathwise estimation is not valid, and we can prove this is the case.

If we compute $\frac{dY}{d\S_0}$ then we know that this value is $0$ everywhere, because the indicator function is a constant value, which means it has no rate of change, **except** for at the point of discontinuity, where the function jumps to a different constant value.

Formally, we use the chain rule:

$\frac{dY}{dS_0} = \frac{dY}{dS_T} \frac{dS_T}{dS_0}$

where $\frac{dY}{dS_T} = 0 $ if $S_T > K$ or $S_T < K$
and $\frac{dY}{dS_T}$ does not exist if $S_T = K$

Thus,

$0 = \mathbb{E}\left( \frac{dY}{dS_0} \right) \neq \frac{d}{dS_0} \mathbb{E} [Y]$

Again, the change in $\mathbb{E}[Y]$ is due to a change in $S_0$ due to the possibility that a change in $S_0$ will cause $S_T$ to cross (or not cross) the barrier $K$. But this change cannot be captured by the pathwise derivative which is zero almost surely.

To code this, it's quite simple. Compute the derivative analytically first, then plug in the values into that expression. No additional simulation is needed. For example, the pathwise delta is given by:

$\frac{dY}{dS_0} = \frac{dY}{dS_T}\frac{dS_T}{dS_0} = e^{-rT} \mathbb{I}_{S_T > K} \frac{S_T}{S_0}$

It should be easy to tell that *this function* is not continuous, and thus its derivative does not exist, meaning the gamma cannot be pathwise-estimated.

In [43]:
##pathwise, using CRN.
np.random.seed(12345)
z1 = np.random.randn(nN)
tmp1  =  s0      *np.exp((r-q-sig**2/2)*T)
tmp2 = sig*np.sqrt(T)

#tmp1 * tmp2
s =  tmp1*np.exp(tmp2*z1)
#s = s0         *np.exp((r-q-sig**2/2)*T + sig*np.sqrt(T)*z1)
dhds0_Call = np.exp(-r*T)*(s>K)* (s)/s0
dhds0_Put = np.exp(-r*T)* (K>s)* (-s)/s0
delC_pathwise = np.mean(dhds0_Call)
delP_pathwise = np.mean(dhds0_Put)

gam_pathwise = np.nan
gam_pathwise = 'undefined' if np.isnan(gam_pathwise) else gam_pathwise

print(f"""
The pathwise estimator for the Delta is:
    Call: {delC_pathwise:.4f}
    Put: {delP_pathwise:.4f}
""")

print(f"""
The pathwise estimator for the Gamma is:
    {gam_pathwise}
""")


The pathwise estimator for the Delta is:
    Call: 0.6250
    Put: -0.3605


The pathwise estimator for the Gamma is:
    undefined



### Likelihood Ratio Method


The likelihood ratio method invovles computing the derivative

$\alpha ^\prime (\theta) = \frac{d}{d\theta}\mathbb{E}[Y(\theta)] $

just as in the pathwise estimation method, however we will rely on the probability density function of the parameter in question. That is to say:


$\alpha ^\prime (\theta) = \frac{d}{d\theta}\mathbb{E}[Y(\theta)] = \frac{d}{d\theta}\mathbb{E}_{\theta}[Y]$

The right hand term means we are computing the expectation with respect to $\theta$. If we assume that $\theta \sim g_{\theta}(x)$, then

$\frac{d}{d\theta}\mathbb{E}_{\theta}[Y] = \frac{d}{d\theta} \int_\mathbb{{R}^d} f(x)g_{\theta}(x)dx$

Now, since integrals are linear, the differentiation operator can be moved inside:

$=  \int_\mathbb{{R}^d}  f(x) \frac{d}{d\theta} g_{\theta}(x)dx$

Now, we rewrite $\frac{dg}{d\theta}$ as $\dot{g}_{\theta} (x)$ which gives us:

$ \int_\mathbb{{R}^d}  f(x)  \dot{g}_{\theta} (x) dx$

But recall we were originally trying to compute an expectation with respect to ${g}_{\theta} (x)$. So, we multiply and divide by this quantity to obtain:

$ \int_\mathbb{{R}^d}  f(x)   \frac{\dot{g}_{\theta}(x)}{g_{\theta}(x)} g_{\theta}(x) dx$

which can be re-expressed as:

$ \mathbb{E}_{\theta}  \left[ f(X) \frac{\dot{g}_{\theta}(X)}{g_{\theta}(X)} \right]$

Note the change from $x$, the integration variable  to $X$, the random variable.

The ratio of $\dot{g}$ to $g$ is known as the score function. This score function can be computed analytically, if the density is known and differentiable, which it is for the case of options following a GBM. With this, we obtain an unbiased estimator of $\alpha ^\prime (\theta) = \frac{1}{n} \sum\limits_{i=1}^{n} f(X_i) \frac{\dot{g_{\theta}}(X_i)}{g_{\theta}(X_i)}$



In [34]:
#s = ST
scoreH_delta_C = np.exp(-r*T)*np.maximum(s - K,0)*z1/(s0*sig*np.sqrt(T))
delC_likelihood = np.mean(scoreH_delta_C)

scoreH_delta_P = np.exp(-r*T)*np.maximum(K - s,0)*z1/(s0*sig*np.sqrt(T))
delP_likelihood = np.mean(scoreH_delta_P)

scoreH_gamma = (z1**2 - 1)/(s0**2*sig**2*T) - z1/(s0**2*sig*np.sqrt(T))
scoreH_gamma *= np.exp(-r*T)*np.maximum(s - K,0)
gam_likelihood = np.mean(scoreH_gamma)


print(f"""
The LRM estimator for the Delta is:
    Call: {delC_likelihood:.4f}
    Put: {delP_likelihood:.4f}
""")

print(f"""
The LRM estimator for the Gamma is:
    Call and Put : {gam_likelihood:.4f}
""")


The LRM estimator for the Delta is:
    Call: 0.6262
    Put: -0.3597


The LRM estimator for the Gamma is:
    Call and Put : 0.0010



Finally, we examine Vega. The process can be extended for other greeks as well.

In [38]:
z2 = np.random.randn(nN)

tmp1U  =  s0 * np.exp((r - q - (sig + delSig)**2/2)*T)
tmp1D  =  s0 * np.exp((r - q - (sig - delSig)**2/2)*T)

tmp2U = (sig + delSig) * np.sqrt(T)
tmp2D = (sig - delSig) * np.sqrt(T)

# (1) W/ Common Random Numbers
s = tmp1U * np.exp(tmp2U*z1)
tmpU = np.exp(-r*T) * np.maximum(s-K,0)

s = tmp1D * np.exp(tmp2D*z1)
tmpD = np.exp(-r*T) * np.maximum(s-K,0)

vega_CRN = np.mean((tmpU - tmpD)/(2*delSig))

# (2) W/O Common Random Numbers
s = tmp1U * np.exp(tmp2U*z1)
tmpU = np.exp(-r*T) * np.maximum(s-K,0)

s = tmp1D * np.exp(tmp2D*z2)
tmpD = np.exp(-r*T) * np.maximum(s-K,0)

vega_tilde = np.mean((tmpU - tmpD)/(2*delSig))

# (3) Pathwise Estimator
s =  tmp1 * np.exp(tmp2*z1)
dhdsig = np.exp(-r*T) * (s>K) * (s/sig) * (np.log(s/s0) - (r - q+sig**2/2)*T)
vega_pathwise = np.mean(dhdsig)

# (4) Likelihood Ratio

#scoreH = exp(-r*T)*np.maximum(s-K,0)*(sig*(sig*T-np.sqrt(T)*z1)-1)/sig
scoreH = np.exp(-r*T)*np.maximum(s - K,0)*((z1**2-1)/sig - z1*np.sqrt(T))
vega_likelihood = np.mean(scoreH)

print('Vega (same for calls and puts)')
print('==================')
print('Exact: %f' % vegaC)
print('W/ CRN: %f' % vega_CRN)
print('W/O CRN: %f' % vega_tilde)
print('Pathwise estimator: %f' % vega_pathwise)
print('Likelihood estimator: %f' % vega_likelihood)

Vega (same for calls and puts)
Exact: 444.834513
W/ CRN: 446.016728
W/O CRN: 596.445724
Pathwise estimator: 446.018121
Likelihood estimator: 447.101050
